### ISSUE TITLE
Elasticsearch index creation keeps looking for "_doc"

### ISSUE THREAD

**Issue by paolodedo** [spring-projects/spring-data-elasticsearch#1654]
SpringBoot 2.4.1
Spring Data Elasticsearch 4.1.2

Bug:

I am trying to create the mapping for my index, through the following code:

```
@Override
    public Mono<Boolean> createIndex() {

        var request = new CreateIndexRequest(getIndex());
        request.source(getMappingSource(), getXContentType());
        return reactiveElasticsearchClient.indices()
            .createIndex(request)
            .doOnSuccess(success ->
                log.info("Index Creation \"{}\": {}", getIndex(), success));
    }
```

and this mapping:

```
public static final String DEFAULT_ES_MAPPING_SOURCE =
		"{\n" +
		"    \"mappings\": {\n" +
		"        \"properties\": {\n" +
		"            \"timestamp\": {\n" +
		"                \"type\": \"date\",\n" +
		"                \"format\": \"epoch_millis||yyyy-MM-dd HH:mm:ss||yyyy-MM-dd\"\n" +
		"            }\n" +
		"        }\n" +
		"    }\n" +
		"}";
```

This error comes out: `org.elasticsearch.ElasticsearchStatusException: Elasticsearch exception [type=mapper_parsing_exception, reason=Failed to parse mapping [_doc]: No type specified for field [properties]], Elasticsearch exception [type=mapper_parsing_exception, reason=No type specified for field [properties]]
`

If I add the "_doc" before "properties" in the mapping JSON (as the previous exception asked me), I get another error (non-sense): `org.elasticsearch.ElasticsearchStatusException: Elasticsearch exception [type=illegal_argument_exception, reason=The mapping definition cannot be nested under a type [_doc] unless include_type_name is set to true.], Elasticsearch exception [type=illegal_argument_exception, reason=The mapping definition cannot be nested under a type [_doc] unless include_type_name is set to true.]
`

Could you please fix it? 

EDIT:

I have found an answer (https://github.com/elastic/elasticsearch/issues/40897#issuecomment-480833056) that shows that your class (ReactiveElasticsearchClient) accepts the wrong **CreateIndexRequest**.

org.elasticsearch.action.admin.indices.create.CreateIndexRequest is compatible with ES 6.x
org.elasticsearch.client.indices.CreateIndexRequest is compatible with ES 7.x [BUT I cannot use it since the client accepts the other]

**Comment by sothawo** [comment]
`ReactiveIndexOperations`has methods for creating indices, writing mappings and settings. What is  missing in these methods that you implement your own `createIndex()`? Is this in a custom `ReactiveIndexOperations` implementation?

Although the code is using `org.elasticsearch.action.admin.indices.create.CreateIndexRequest` it's working up to the actual Elasticsearch version 7.10.2.

**Comment by paolodedo** [comment]
Do not know where you are looking but.....no, there are not.

I'd just like to create an index with a specific mapping, feature not present in SDE atm.

No, CreateIndexRequest is not accepted.

**Comment by sothawo** [comment]
There is `org.springframework.data.elasticsearch.core.ReactiveIndexOperations#create()` and `org.springframework.data.elasticsearch.core.ReactiveIndexOperations#putMapping()` and these functions definitely work. 

But we should indeed use the same `CreateIndexRequest` as we do in the non-reactive code. The same for the `GetIndexRequest`, `PutMappingRequest` and `GetMappingRequest`. I created #1658 to address that.

**Comment by paolodedo** [comment]
How about passing the body to that putMapping, as with zero parameters is quite unuseful.

**Comment by paolodedo** [comment]
Still, another kind of error:

applying the previous JSON

```
var putMappingRequest = new PutMappingRequest(getIndex());
putMappingRequest.source(getMappingSource(), getXContentType());

return reactiveElasticsearchClient.indices()
    .putMapping(putMappingRequest)
    .doOnSuccess(success ->
        log.info("Index Mapping \"{}\": {}", getIndex(), success));
```

comes out `org.elasticsearch.ElasticsearchStatusException: Elasticsearch exception [type=mapper_parsing_exception, reason=Root mapping definition has unsupported parameters:  [mappings : {properties={timestamp={format=epoch_millis||yyyy-MM-dd HH:mm:ss||yyyy-MM-dd, type=date}}}]], Elasticsearch exception [type=mapper_parsing_exception, reason=Root mapping definition has unsupported parameters:  [mappings : {properties={timestamp={format=epoch_millis||yyyy-MM-dd HH:mm:ss||yyyy-MM-dd, type=date}}}]]
`

**Comment by paolodedo** [comment]
Exact same issue using putMapping(Mono<Document> mapping)

**Comment by sothawo** [comment]
as I already wrote it create index, get idnex, put mapping, get mapping; that's what I created the other issue for.

> How about passing the body to that putMapping, as with zero parameters is quite unuseful

I assume you only use the `ReactiveElasticsearchClient` from Spring Data Elasticsearch? Because the `putMapping()` does make sense for the `IndexOperations` when they are created for an entity class. Spring Data Elasticsearch will use this entity class, and create a mapping according to the properties of the entity that are annotated with `@Field`.

**Comment by paolodedo** [comment]
I am not using mapped entities, though.

**Comment by paolodedo** [comment]
In order to workaround this, is there the possibility of passing _include_type_name=true_ on my `org.elasticsearch.action.admin.indices.create.CreateIndexRequest` ?

**Comment by sothawo** [comment]
Closing this ticket, as the issue is addressed in #1658.
